---
bool end_of_stream